File: prd-api/src/PrdAgent.Core/Models/TaskTree.cs
namespace PrdAgent.Core.Models;

/// <summary>
/// 个人任务树 — 以"创世支柱 → 主干 → 枝干"的方式组织个人任务。
/// 每个用户可拥有多棵树；树根节点即"创世支柱"，向下生长出主干与枝干。
/// 目标：方便地（含对话摘取）把个人任务及其卡点暴露出来，让自己和上级一眼看清推进进度。
/// </summary>
public class TaskTree
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");

    /// <summary>树标题（创世支柱标题，由用户输入或对话提取）</summary>
    public string Title { get; set; } = string.Empty;

    public string? Description { get; set; }

    /// <summary>所属用户</summary>
    public string OwnerId { get; set; } = string.Empty;

    /// <summary>节点总数（反规范化缓存）</summary>
    public int NodeCount { get; set; }

    /// <summary>树的最大深度（反规范化缓存）</summary>
    public int MaxDepth { get; set; }

    /// <summary>是否归档（归档后不在默认列表展示）</summary>
    public bool IsArchived { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

File: prd-admin/src/services/contracts/taskTree.ts
import type { ApiResponse } from '@/types/api';

// ── 数据类型 ──

export type TaskStatus = 'idea' | 'planned' | 'building' | 'done' | 'blocked';

export type TaskTree = {
  id: string;
  title: string;
  description?: string;
  ownerId: string;
  nodeCount: number;
  maxDepth: number;
  isArchived: boolean;
  createdAt: string;
  updatedAt: string;
};

export type TaskNode = {
  id: string;
  treeId: string;
  ownerId: string;
  parentId?: string | null;
  dependsOn: string[];
  title: string;
  description?: string;
  status: TaskStatus;
  blocker?: string | null;
  blockedSince?: string | null;
  order: number;
  positionX: number;
  positionY: number;
  tags: string[];
  createdAt: string;
  updatedAt: string;
};

/** 卡点上报项 */
export type TaskBlockerItem = {
  node: TaskNode;
  ownerName: string;
  treeTitle: string;
  stuckDays: number;
  blocks: string[];
};

// ── 请求类型 ──

export type CreateTaskTreeInput = {
  title: string;
  description?: string;
  rootStatus?: TaskStatus;
};

export type CreateTaskNodeInput = {
  parentId?: string | null;
  title: string;
  description?: string;
  status?: TaskStatus;
  blocker?: string;
  dependsOn?: string[];
  order?: number;
};

export type UpdateTaskNodeInput = {
  title?: string;
  description?: string;
  status?: TaskStatus;
  blocker?: string;
  parentId?: string | null;
  order?: number;
  positionX?: number;
  positionY?: number;
};

// ── Contract 签名 ──

export type CreateTaskTreeContract = (
  input: CreateTaskTreeInput
) => Promise<ApiResponse<{ tree: TaskTree; root: TaskNode }>>;

export type ListTaskTreesContract = (
  includeArchived?: boolean
) => Promise<ApiResponse<{ items: TaskTree[]; total: number }>>;

export type GetTaskTreeContract = (
  treeId: string
) => Promise<ApiResponse<{ tree: TaskTree; nodes: TaskNode[] }>>;

export type DeleteTaskTreeContract = (
  treeId: string
) => Promise<ApiResponse<{ deleted: boolean }>>;

export type CreateTaskNodeContract = (
  treeId: string,
  input: CreateTaskNodeInput
) => Promise<ApiResponse<TaskNode>>;

export type UpdateTaskNodeContract = (
  nodeId: string,
  input: UpdateTaskNodeInput
) => Promise<ApiResponse<TaskNode>>;

export type DeleteTaskNodeContract = (
  nodeId: string
) => Promise<ApiResponse<{ deleted: number }>>;

export type AddTaskDependencyContract = (
  nodeId: string,
  dependsOnId: string
) => Promise<ApiResponse<{ added: boolean }>>;

export type RemoveTaskDependencyContract = (
  nodeId: string,
  dependsOnId: string
) => Promise<ApiResponse<{ removed: boolean }>>;

export type ListTaskBlockersContract = (
  scope?: 'mine' | 'all'
) => Promise<
  ApiResponse<{ items: TaskBlockerItem[]; total: number; canViewAll: boolean; scope: string }>
>;
